File: ios/CourtSearchKit/Sources/CourtSearchKit/Embedder.swift
import CourtEmbedder
import Foundation

/// Text to unit vector, on the device: the sample's own Rust embedder
/// (embedder-ffi) over a Model2Vec table. The engine ships no embedder by
/// design, so the app is the caller that supplies vectors.
final class Embedder {
    private let handle: OpaquePointer
    let dimensions: Int

    init(directory: URL) throws {
        guard let handle = court_embedder_open(directory.path) else {
            throw EngineError(operation: "embedder", code: -1, message: String(cString: court_embedder_last_error()))
        }
        self.handle = handle
        dimensions = court_embedder_dim(handle)
    }

    deinit { court_embedder_close(handle) }

    /// WordPiece pieces the text tokenizes to, `[UNK]`s included.
    func pieces(_ text: String) -> Int {
        text.utf8CString.withUnsafeBufferPointer { bytes in
            bytes.withMemoryRebound(to: UInt8.self) { court_embedder_pieces(handle, $0.baseAddress, $0.count - 1) }
        }
    }

    /// Nil when the text has no vector: empty, or entirely out of vocabulary.
    func embed(_ text: String) throws -> [Float]? {
        var vector = [Float](repeating: 0, count: dimensions)
        let written = text.utf8CString.withUnsafeBufferPointer { bytes in
            bytes.withMemoryRebound(to: UInt8.self) { utf8 in
                // utf8CString carries a trailing NUL the embedder must not see.
                court_embedder_embed(handle, utf8.baseAddress, utf8.count - 1, &vector, vector.count)
            }
        }
        if written < 0 {
            throw EngineError(operation: "embed", code: -1, message: String(cString: court_embedder_last_error()))
        }
        return written == 0 ? nil : vector
    }
}

File: android/app/src/main/java/ai/pipestream/samples/courtsearch/Embedder.kt
package ai.pipestream.samples.courtsearch

import android.content.res.AssetManager
import java.io.File

/**
 * Text to unit vector, on the device: the sample's own Rust embedder
 * (embedder-ffi) over a Model2Vec table. The engine ships no embedder by design,
 * so the app is the caller that supplies vectors.
 */
class Embedder(directory: File) : AutoCloseable {
    private val handle: Long = nativeOpen(directory.path)
    val dimensions: Int

    init {
        if (handle == 0L) throw EngineException("embedder", -1, nativeLastError())
        dimensions = nativeDim(handle)
    }

    /** WordPiece pieces the text tokenizes to, `[UNK]`s included. */
    fun pieces(text: String): Int = nativePieces(handle, text)

    /** Null when the text has no vector: empty, or entirely out of vocabulary. */
    fun embed(text: String): FloatArray? = nativeEmbed(handle, text)

    override fun close() = nativeClose(handle)

    companion object {
        const val MODEL = "potion-retrieval-32M"

        init {
            System.loadLibrary("court_embedder_ffi")
        }

        /**
         * The embedder maps the table from a file, and APK assets are not files, so
         * the model is copied into app storage once. Returns null when the APK was
         * built without the model (scripts/fetch-model.sh): search by meaning is
         * then simply not offered.
         */
        fun install(assets: AssetManager, filesDir: File): File? {
            val names = listOf("tokenizer.json", "model.safetensors")
            if (assets.list(MODEL).orEmpty().toSet().containsAll(names).not()) return null
            val target = File(filesDir, "models/$MODEL").apply { mkdirs() }
            for (name in names) {
                val out = File(target, name)
                val length = assets.openFd("$MODEL/$name").use { it.length }
                if (out.length() == length) continue
                assets.open("$MODEL/$name").use { input -> out.outputStream().use { input.copyTo(it, 1 shl 20) } }
            }
            return target
        }

        @JvmStatic private external fun nativeOpen(directory: String): Long
        @JvmStatic private external fun nativeDim(handle: Long): Int
        @JvmStatic private external fun nativeEmbed(handle: Long, text: String): FloatArray?
        @JvmStatic private external fun nativePieces(handle: Long, text: String): Int
        @JvmStatic private external fun nativeClose(handle: Long)
        @JvmStatic private external fun nativeLastError(): String
    }
}
